### PR TITLE
add halo updating to convolutional training and prediction

### PR DESCRIPTION
**Pull request by mcgibbon**
This PR updates convolutional training and Predictor behavior such that convolutional layers no longer zero-pad, and instead the input data has halo points appended prior to training and prediction. Halos are updated only on tile edges, not corners where there is no "correct" tile data. This change means convolutional training and (currently) convolutional Predictor prediction additionally requires a "tile" dimension in its input data, unless MPI is present and running on at least 6 ranks in a square-layout fashion.

Refactored public API:
- Convolutional nets now use halo-padding instead of zero-padding, and require a "tile" dimension to be present when training
- during prediction, convolutional predictors require a "tile" dimension unless running with MPI on a square layout

Significant internal changes:
- Added function `append_halos` which takes in an xarray dataset and a number of halo points and appends halos to those points.
- Refactored private methods in `fv3fit._shared.stacking` to be package-private instead of module-private, in order to use them directly within XyMultiArraySequence.
- Refactored XyMultiArraySequence and Predictor to take in a `n_halo` argument indicating how many halo points the X data needs appended.
- Refactored XyMultiArraySequence to take on the responsibilities of StackedBatches, including stacking, shuffling, dropping nans, and calls to `check_empty` and `preserve_samples_per_batch`.
- Refactored `shuffled` helper function to take in a sequence of datasets with identical coordinates, and shuffle them identically.
- Renamed XyArraySequence, ThreadedSequencePreloader, and XyMultiArraySequence to remove the leading underscore since they are package-private and not module-private.
- Refactored precipitative and convolutional training codes to stack within XyMultiArraySequence instead of stacking beforehand using StackedBatches, since stacking must occur after halo updating and halo updating must only occur on X data (not y data).
- Updated training tests to include a tile dimension, while keeping total number of sample points somewhat consistent.

Requirements changes:
- fv3fit now depends on fv3gfs-util to perform halo updates

- [x] Tests added

Before appending halos:
Figure_1
After appending halos:
Figure_2

Resolves #<github issues> [JIRA-TAG]

(Delete this for the commit message)
You are encouraged to check the PR against the [Code Review Checklist](https://paper.dropbox.com/doc/Code-Review-Checklist--A4lKrs~xg7w5Gsb39N6JLNQoAg-IlsYffZgTwyKEylty7NhY).

### CODE REVIEW

File: external/fv3fit/examples/append_halo.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import xarray as xr
```
Comment (mcgibbon) [inline external/fv3fit/examples/append_halo.py:1]: I made this to test out the halo update code, could delete it though I thought I'd propose adding it as an example. Feels kind of strange as an example since this is private API within fv3fit, so I could go either way.

File: external/fv3fit/fv3fit/keras/_models/shared/__init__.py [hunk 2]
```diff
@@ -15,3 +15,5 @@
     standard_normalize,
     count_features,
 )
+from .sequences import XyArraySequence, XyMultiArraySequence, ThreadedSequencePreLoader
+from .halos import append_halos, HalosBatches
```
Comment (mcgibbon) [inline external/fv3fit/fv3fit/keras/_models/shared/__init__.py:19]: I never ended up using HalosBatches, I think I will remove it. In general it is probably never useful because you don't want output data to have halos.

File: external/fv3fit/fv3fit/keras/_models/shared/halos.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    datasets_in = [ds.isel(tile=i) for i in range(0, 6)]
+    quantity_dicts = [_quantities_from_dataset(ds, n_halo=n_halo) for ds in datasets_in]
+    for name in ds.data_vars.keys():
+        quantities = [d[name] for d in quantity_dicts]
+        if n_halo > 0:
+            _halo_update(communicators, quantities, n_halo=n_halo)
+    datasets_out = [_dataset_from_quantities(d) for d in quantity_dicts]
+    ds_out = xr.concat(datasets_out, dim="tile")
+    ds_out = match_prediction_to_input_coords(input=ds, prediction=ds_out)
+    # if you don't drop coords here, the model fails to train - no idea why
```
Comment (mcgibbon) [inline external/fv3fit/fv3fit/keras/_models/shared/halos.py:138]: One of the grand mysteries of xarray. `xr.testing.assert_identical(ds_out, ds_out.drop(ds_out.coords.keys())` does not raise AssertionError, but if I return the first one (`ds_out`) the model fails to train, staying at around 0.97 loss. Printing `ds_out` shows no coordinates present. Printing `ds_out.coords.keys()` shows a KeysView that says coordinates are empty.

File: external/fv3fit/tests/keras/test_convolutional_network.py [hunk 4]
```diff
@@ -54,25 +55,37 @@ def test_standard_input_gives_standard_output():
     assert out_std > 0.1
 
 
+def get_X_y(ds, X_names, y_names):
+    return tuple(ds[n].values for n in X_names), tuple(ds[n].values for n in y_names)
```
Comment (mcgibbon) [inline external/fv3fit/tests/keras/test_convolutional_network.py:59]: I didn't end up using this function and should delete it.

**Review by AnnaKwa — CHANGES_REQUESTED** [review]
Thanks, this is neat to have this feature in the training now. I had a few minor comments. However, I don't have a lot of experience with the halo updates using the communicators, so while I did review `halos.py` I am mostly just trusting that it is the same implementation that's used in the dycore and I'm not 100% confident in my ability to catch issues there.

Just to confirm, the default kernel size in the hyperparameters is such that the halo appending is actually getting tested in the fit and predict methods during `test_train.py`, right?

**Comment by mcgibbon** [comment]
> However, I don't have a lot of experience with the halo updates using the communicators, so while I did review halos.py I am mostly just trusting that it is the same implementation that's used in the dycore and I'm not 100% confident in my ability to catch issues there.

Do you think the unit tests I included are enough to ensure this is working? I tried to avoid including thorough unit tests of what values are correct and just include a basic test because the halo updates are tested thoroughly in fv3gfs-util. Was this the right call?

> Just to confirm, the default kernel size in the hyperparameters is such that the halo appending is actually getting tested in the fit and predict methods during test_train.py, right?

Yes, the default kernel size of 3 requires halos.

**Review by brianhenn — COMMENTED** [review]
Hi, I started reviewing this but haven't gotten all of the way through it. Wanted to support some of Anna's comments by responding now, but will keep at it.

File: external/fv3fit/tests/keras/test_halos.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+                n_halo=n_halo,
+                comm=comms[i],
+            )
+        )
+    output_datasets[0] = _append_halos_using_mpi(
+        ds=rank_datasets[0], n_halo=n_halo, comm=comms[0]
+    )
+    for i in range(n_tile):
+        check_result(result=output_datasets[i], n_halo=n_halo, nx=nx, ny=ny, nz=nz)
+        xr.testing.assert_identical(non_mpi_result.isel(tile=i), output_datasets[i])
```
Comment (brianhenn) [inline external/fv3fit/tests/keras/test_halos.py:159]: Could just `rank_datasets[i]` instead of non_mpi_result

File: external/fv3fit/tests/keras/test_halos.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        )
+
+    return callback
+
+
+@pytest.mark.parametrize(
+    "nx, ny, nz, n_halo",
+    [pytest.param(8, 8, 12, 3, id="typical"), pytest.param(8, 8, 12, 1, id="one_halo")],
+)
+def test_with_and_without_mpi_give_same_result(nx: int, ny: int, nz: int, n_halo: int):
```
Comment (brianhenn) [inline external/fv3fit/tests/keras/test_halos.py:129]: Clever idea to test this way. Just to make sure I understand, the test adds a daisy-chain blocking sequence to the mpi-like case to simulate real MPI async behavior despite the fact that it's actually using a dummy com? And `Recv` is triggered by the halo update? Because outside of the comm the mpi- and non-mpi halo updates aren't really different so I can see that this is kind of hard to test.

**Review by brianhenn — COMMENTED** [review]
Thanks @mcgibbon this was fun to see how it was implemented. My main request is some cleanup of `build_model` and in particular the normalize/denormalize parts, which I found messy and hard to follow. Otherwise this looks good.

File: external/fv3fit/fv3fit/keras/_models/convolutional.py [hunk 7]
```diff
@@ -85,41 +82,47 @@ def train_convolutional_model(
     train_batches: Sequence[xr.Dataset],
     validation_batches: Optional[Sequence[xr.Dataset]] = None,
 ):
+    n_halo = hyperparameters.convolutional_network.halos_required
     if validation_batches is not None:
-        validation_data: Optional[
-            Tuple[Tuple[Any, ...], Tuple[Any, ...]]
-        ] = batch_to_array_tuple(
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/convolutional.py:91]: Deleted.

File: external/fv3fit/fv3fit/keras/_models/convolutional.py [hunk 8]
```diff
@@ -133,24 +136,40 @@ def batch_to_array_tuple(
...
+def _non_vertical_dims(dims: Sequence[str]):
+    return [entry for entry in dims if entry not in ("z", "z_interface")]
+
+
 def build_model(
-    config: ConvolutionalHyperparameters, batch: xr.Dataset
+    config: ConvolutionalHyperparameters,
+    X: Sequence[np.ndarray],
+    y: Sequence[np.ndarray],
+    batch: xr.Dataset,
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/convolutional.py:158]: I'll look at switching this to take either X/y or batch. I seem to remember there being a reason X/y weren't sufficient but maybe not.

File: external/fv3fit/fv3fit/keras/_models/convolutional.py [hunk 9]
```diff
@@ -133,24 +136,40 @@ def batch_to_array_tuple(
...
+        config: configuration of convolutional training
+        X: example input for keras fitting, used to determine shape and normalization
+        y: example output for keras fitting, used to determine shape and normalization
+        batch: data from X and y in the form of an xarray dataset
+    """
+    input_layers = [tf.keras.layers.Input(shape=array.shape[1:]) for array in X]
     norm_input_layers = standard_normalize(
         names=config.input_variables,
         layers=input_layers,
         batch=batch,
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/convolutional.py:171]: I'll check it out.

File: external/fv3fit/fv3fit/keras/_models/shared/halos.py [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+    def __len__(self) -> int:
+        return len(self._batches)
+
+
+def _create_comms(total_ranks: int) -> Sequence[fv3gfs.util.testing.DummyComm]:
+    buffer_dict: Dict[Any, Any] = {}
+    comms: List[fv3gfs.util.testing.DummyComm] = []
+    for rank in range(total_ranks):
+        comms.append(
+            fv3gfs.util.testing.DummyComm(
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/shared/halos.py:13]: DummyComm should probably be renamed DictComm and moved to a non-testing namespace. It only hasn't been because that is a backwards-compatibility-breaking change. We've always had planned that DummyComm would enable non-MPI communication like this, and its features are unit tested.

File: external/fv3fit/fv3fit/keras/_models/shared/halos.py [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+    n_tiles = ds.sizes["tile"]
+    if n_tiles != 6:
+        raise ValueError(f"dataset must have 6 tiles to halo update, got {n_tiles}")
+    comms = _create_comms(total_ranks=6)
+    communicators = [_get_cubed_sphere_communicator(comm) for comm in comms]
+    datasets_in = [ds.isel(tile=i) for i in range(0, 6)]
+    quantity_dicts = [_quantities_from_dataset(ds, n_halo=n_halo) for ds in datasets_in]
+    for name in ds.data_vars.keys():
+        quantities = [d[name] for d in quantity_dicts]
+        if n_halo > 0:
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/shared/halos.py:149]: Simplifying logic. Because of this, I can have a line that says "make sure this dataset has as many halo lines as it needs" instead of saying "if this dataset needs at least one halo line, make sure this dataset has as many halo lines as it needs" which is more complex.

It also makes the code behavior more congruous between the with-halo and no-halo workflows, without this I was getting the esoteric xarray failure only when halos are being used.

File: external/fv3fit/fv3fit/keras/_models/shared/pure_keras.py [hunk 12]
```diff
@@ -44,12 +46,14 @@ def __init__(
             model: keras model to wrap
             unstacked_dims: if given, stacking should leave these dimensions in place.
                 by default uses z-dimension names from fv3gfs.util
+            n_halo: if given, number of halo points required in input data
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/shared/pure_keras.py:49]: Removed "if given", since this is always given and the default is written in the keyword argument. Should only need to state the true default in the docstring when the kwarg says None.

File: external/fv3fit/fv3fit/keras/_models/shared/sequences.py [hunk 13]
```diff
@@ -37,37 +47,55 @@ def __getitem__(self, idx) -> Tuple[np.ndarray, np.ndarray]:
...
+        self.unstacked_dims = unstacked_dims
 
     def __len__(self) -> int:
         return len(self.dataset_sequence)
 
     def __getitem__(self, idx) -> Tuple[np.ndarray, np.ndarray]:
         ds = self.dataset_sequence[idx]
-        X = tuple(ds[name].values for name in self.X_names)
-        y = tuple(ds[name].values for name in self.y_names)
+        X_y_datasets = [
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/shared/sequences.py:100]: I hopefully made it a little cleaner by at least not doing the X halo update using the list container. The issue I run into when doing that is you can't have a for loop to change the value of a local variable in-place, unless it's contained in a mutable type like a dict or list (or if I do something overly fancy with `locals()`).

File: external/fv3fit/fv3fit/keras/_models/shared/sequences.py [hunk 14]
```diff
@@ -37,37 +46,52 @@ def __getitem__(self, idx) -> Tuple[np.ndarray, np.ndarray]:
...
-        y = tuple(ds[name].values for name in self.y_names)
+        X_y_datasets = [ds[self.X_names], ds[self.y_names]]
+        X_y_datasets[0] = append_halos(X_y_datasets[0], n_halo=self.n_halo)
+        for i in range(2):
+            X_y_datasets[i] = stack(
+                X_y_datasets[i], unstacked_dims=self.unstacked_dims
+            ).dropna(dim=SAMPLE_DIM_NAME)
+            X_y_datasets[i] = check_empty(X_y_datasets[i])
+            X_y_datasets[i] = preserve_samples_per_batch(X_y_datasets[i])
+        X_y_datasets = shuffled(np.random, X_y_datasets)
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/shared/sequences.py:111]: I also can't call such a helper separately on X and y because they need to be shuffled identically, without doing something risky like passing an identical random seed. If a bug ever makes these non-identical, it would be very hard to track down.

File: external/fv3fit/tests/keras/test_halos.py [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+    [
+        pytest.param(8, 8, 12, 6, 3, id="typical"),
+        pytest.param(8, 8, 12, 6, 1, id="one_halo"),
+        pytest.param(8, 8, 12, 6, 0, id="no_halo"),
+    ],
+)
+def test_append_halos_extends_dims(nx: int, ny: int, nz: int, n_tile: int, n_halo: int):
+    ds = get_dataset(nx=nx, ny=ny, nz=nz, n_tile=n_tile)
+    for name, da in ds.data_vars.items():
+        assert da.sizes["x"] == nx, name
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/tests/keras/test_halos.py:45]: Later I check that the size of X is nx + 2 * n_halo. If the starting size isn't nx (which it should be unless the get_dataset helper has a bug), it will cause the test to fail. If there's a bug in get_dataset, I'd like it to show up here so we know the bug is in the test infrastructure and not the code.

File: external/fv3fit/tests/keras/test_halos.py [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+    check_result(result=result, n_halo=n_halo, nx=nx, ny=ny, nz=nz)
+    assert result.dims.keys() == ds.dims.keys()
+
+
+def check_result(result: xr.Dataset, n_halo: int, nx: int, ny: int, nz: int):
+    for name, da in result.data_vars.items():
+        assert da.sizes["x"] == nx + 2 * n_halo, name
+        assert da.sizes["y"] == ny + 2 * n_halo, name
+        assert da.sizes.get("z", nz) == nz, name
+        # only corners should still be zero
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/tests/keras/test_halos.py:58]: It's helpful to know if the code is so wrong that it's overwriting compute data, versus if it is failing to write to the halo.

File: external/fv3fit/tests/keras/test_halos.py [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+        # x and y edge halos should be 1 now, but corners are still zero
+        data_with_x_halos = da.isel(
+            x=range(0, nx + 2 * n_halo), y=range(n_halo, n_halo + ny)
+        )
+        assert np.sum(data_with_x_halos == 0) == 0, name
+        data_with_y_halos = da.isel(
+            x=range(n_halo, n_halo + nx), y=range(0, ny + 2 * n_halo)
+        )
+        assert np.sum(data_with_y_halos == 0) == 0, name
+
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/tests/keras/test_halos.py:73]: Added.

File: external/fv3fit/tests/keras/test_halos.py (hunk 6 above)
Comment (mcgibbon) [reply external/fv3fit/tests/keras/test_halos.py:129]: Yesss, exactly. Each rank will run until it needs to receive data then tell the next one to run completely before it continues. It doesn't work if there's more than 1 send-receive pairs. In the long run we'll hopefully have a solution that uses threading and a shared dictionary, where the thread can wait for other ranks/threads.

File: external/fv3fit/tests/keras/test_halos.py (hunk 5 above)
Comment (mcgibbon) [reply external/fv3fit/tests/keras/test_halos.py:159]: I tried then realized I actually can't. I modified the code above to compute rank_datasets before it computes non_mpi_result to make this clearer. rank_datasets is input data, while non_mpi_result has halos appended.

File: external/fv3fit/tests/training/test_train.py [hunk 18]
```diff
@@ -176,34 +177,43 @@ def test_train_default_model_on_identity(model_type, regtest):
...
     transpose_output["var_out"].values[:] = np.transpose(
-        transpose_output["var_out"].values, axes=(0, 2, 1, 3)
+        transpose_output["var_out"].values, axes=(0, 1, 3, 2, 4)
     )
     output_from_transpose = result.model.predict(transpose_input)
-    xr.testing.assert_allclose(output_from_transpose, transpose_output, atol=1e-5)
+    n_halo = result.hyperparameters.convolutional_network.halos_required
+    # transposing tile data messes up neighbors, so we have to assess only on
+    # data that has no halo dependence
+    assert n_halo * 2 < nx
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/tests/training/test_train.py:210]: If this assertion fails, it means the assertions below will be testing conditions on empty arrays, which can pass trivially even if the code has bugs.

File: external/fv3fit/fv3fit/keras/_models/convolutional.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/convolutional.py]: Refactored. All this was done to ensure x and y were 5d when being fit, so I just made them 5d when being fit.

File: external/fv3fit/fv3fit/_shared/stacking.py [hunk 19]
```diff
@@ -100,23 +100,24 @@ def _preserve_samples_per_batch(ds: xr.Dataset) -> xr.Dataset:
...
-    chunks = dataset.chunks.get(SAMPLE_DIM_NAME, chunks_default)
+    chunks_default = (len(datasets[0][SAMPLE_DIM_NAME]),)
+    chunks = datasets[0].chunks.get(SAMPLE_DIM_NAME, chunks_default)
     chunk_indices = _get_chunk_indices(chunks)
     shuffled_inds = np.concatenate(
         [random.permutation(indices) for indices in chunk_indices]
     )
 
-    return dataset.isel({SAMPLE_DIM_NAME: shuffled_inds})
+    return [dataset.isel({SAMPLE_DIM_NAME: shuffled_inds}) for dataset in datasets]
```
[the remark this answers is not in the document]
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/_shared/stacking.py:122]: Ok I see, I read this first before reading through the `XyMultiArraySequence` so I thought you were shuffing [X0, X1, ...] not [X0, y0].

File: external/fv3fit/fv3fit/keras/_models/shared/sequences.py [hunk 20]
```diff
@@ -37,37 +47,74 @@ def __getitem__(self, idx) -> Tuple[np.ndarray, np.ndarray]:
...
-        X = tuple(ds[name].values for name in self.X_names)
-        y = tuple(ds[name].values for name in self.y_names)
+        X_ds = vcm.safe.get_variables(ds=ds, variables=self.X_names)
+        X_ds = append_halos(X_ds, n_halo=self.n_halo)
+        y_ds = vcm.safe.get_variables(ds=ds, variables=self.y_names)
+        X_y_datasets = [X_ds, y_ds]
+        for i in range(2):
+            X_y_datasets[i] = stack(
+                X_y_datasets[i], unstacked_dims=self.unstacked_dims
+            ).dropna(dim=SAMPLE_DIM_NAME)
```
Comment (AnnaKwa) [inline external/fv3fit/fv3fit/keras/_models/shared/sequences.py:107]: Now that there's >1 dataset that's supposed to get the same indices sampled, if there's somehow a nan entry for some column feature in one dataset but not the other I think that will throw off the indexes after that sample so they don't correspond to the same columns.

**Comment by AnnaKwa** [comment]
> Do you think the unit tests I included are enough to ensure this is working? I tried to avoid including thorough unit tests of what values are correct and just include a basic test because the halo updates are tested thoroughly in fv3gfs-util. Was this the right call?

I think the unit tests cover the fv3fit usage well and that it's fine to trust the fv3gfs-util tests that the updates are getting the correct values.

**Review by AnnaKwa — APPROVED** [review]
Thanks for responding to the comments/questions, LGTM! I'm approving. I did have one question about potential for the NaN-dropping in `XyMultiArraySequence` to get indices misaligned across the datasets being stacked.

**Review by brianhenn — COMMENTED** [review]
Thanks @mcgibbon for the nice `build_model` cleanup. I have a couple minor tidying up comments.

File: external/fv3fit/fv3fit/keras/_models/convolutional.py [hunk 21]
```diff
@@ -147,15 +136,19 @@ def _count_array_features(arr: np.ndarray) -> int:
...
 
 
 def build_model(
     config: ConvolutionalHyperparameters,
     X: Sequence[np.ndarray],
     y: Sequence[np.ndarray],
-    batch: xr.Dataset,
 ) -> Tuple[tf.keras.Model, tf.keras.Model]:
     """
     Args:
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/convolutional.py:154]: Updated.

File: external/fv3fit/fv3fit/keras/_models/shared/sequences.py (hunk 20 above)
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/shared/sequences.py:107]: Instead of dropping nans, I modified this to raise ValueError if nans are present. This (having NaN in our data) shouldn't still happen, but we can always change it back if it does.

File: external/fv3fit/fv3fit/keras/_models/shared/utils.py [hunk 22]
```diff
@@ -57,66 +58,47 @@ def _fit_norm_layer(
     cls: Type[NormLayer],
     names: Sequence[str],
     layers: Sequence[tf.Tensor],
-    batch: xr.Dataset,
-    sample_dims: Sequence[str] = (SAMPLE_DIM_NAME,),
+    arrays: Sequence[np.ndarray],
 ) -> Sequence[NormLayer]:
     out: List[NormLayer] = []
-    input_features = count_features(names=names, batch=batch, sample_dims=sample_dims)
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/shared/utils.py:64]: Deleted.
